webhook/webhook_test.go

```diff
@@ -0,0 +1,158 @@
+package webhook_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/haveachin/infrared/webhook"
+	"net/http"
+	"testing"
+	"time"
+)
+
+type mockHTTPClient struct {
+	*testing.T
+	method string
+	url    string
+	body   *bytes.Buffer
+}
+
+func (mock *mockHTTPClient) Do(req *http.Request) (*http.Response, error) {
+	if req.Method != mock.method {
+		mock.Fail()
+	}
+
+	if req.URL.String() != mock.url {
+		mock.Fail()
+	}
+
+	_, err := mock.body.ReadFrom(req.Body)
+	if err != nil {
+		mock.Error(err)
+	}
+
+	return nil, nil
+}
+
+func TestWebhook_Serve(t *testing.T) {
+	tt := []struct {
+		webhook webhook.Webhook
+		event   webhook.Event
+	}{
+		{
+			webhook: webhook.Webhook{
+				URL:    "https://example.com",
+				Events: []string{webhook.EventTypeError},
+			},
+			event: webhook.ErrorEvent{
+				Error:    "my error message",
+				ProxyUID: "example.com@1.2.3.4:25565",
+			},
+		},
+		{
+			webhook: webhook.Webhook{
+				URL:    "https://example.com",
+				Events: []string{webhook.EventTypePlayerJoin, webhook.EventTypePlayerLeave},
+			},
+			event: webhook.PlayerJoinEvent{
+				Username:      "notch",
+				RemoteAddress: "1.2.3.4",
+				TargetAddress: "1.2.3.4",
+				ProxyUID:      "example.com@1.2.3.4:25565",
+			},
+		},
+	}
+
+	for _, tc := range tt {
+		body := bytes.Buffer{}
+		tc.webhook.HTTPClient = &mockHTTPClient{
+			T:      t,
+			method: http.MethodPost,
+			url:    tc.webhook.URL,
+			body:   &body,
+		}
+
+		eventCh := make(chan webhook.Event)
+		eventLogCh := make(chan *webhook.EventLog)
+		hasStopped := make(chan bool, 1)
+		go func() {
+			if err := tc.webhook.Serve(eventCh, eventLogCh); err != nil {
+				t.Error(err)
+			}
+			hasStopped <- true
+		}()
+		eventCh <- tc.event
+		eventLog := <-eventLogCh
+
+		bb, err := json.Marshal(eventLog)
+		if err != nil {
+			t.Error(err)
+		}
```

Comment:
You can remove the error check here. It doesnt add any additional value. Unless its possible for the code which is being test to create an eventLog which has an invalid json format but since were using a library package for it, we need to trust that it works correctly

Reply:
This is need to ensure that the body that the http client sends is the same that we expect

Reply:
My origin point was that the if statement could be removed, if an error is return `bb` will have a value of nil. if `bb` isnt nil the `err` value will be nil. So there is no point in checking whether or not the err is nil since it check will make it fail anyway, right?

But looking at it more closely what youre testing here is or `json.Marshal(eventLog)` is consistent with its result if you give it the same EventLog struct and or `bytes.NewReader(bb)` doesnt change byte values. Which both of them should already be covered by their own packages.

For example if for some odd reason the EventLog creation code would be like the following: 
``` 
eventLog := EventLog{
		EventType:  "test",
		Timestamp:  time.Time{},
		Event:          EventError{},	
}
```
We would have no idea that this method is faulty unless someone notices it because the test wont fail.

Reply:
So what you are saying is that we should check the other way around? Like that the should unmarshal the EventLog from the mock and check if the EventLogs are equal by value?

I'm thinking about the DispatchEvent method like a black box that needs to be tested for it's functionality. So we can't trust that it doesn't manipulate the request body in some weird way before it gets posted to the HTTP server.

Reply:
Black box is certainly a way of testing code, but in this case it doesnt really add any value at least with this code.
Since you removed Server and Stop, maybe you also wanna remove the return types of the DispatchEvent method? Not sure what youre plan exactly is with it. 

An example the current way of testing would be similar to this:
```
func add(a, b int) int {
  return a + b
} 

func TestAdd(t *testing.T) {
  result := add(1, 2)
  if 1 + 2 != result {
     t.Fail()
  }
}
```

While it would be more useful to have something like this:
```
func add(a, b int) int {
  return a + b
} 

func TestAdd(t *testing.T) {
  expectedResult := 3
  result := add(1, 2)
  if expectedResult != result {
     t.Fail()
  }
}
```
In here we test or the method does its works without using the same way of doing something as we do in the method. If the method `add` where to change and for example go had weird bug which causes + to behave like a * if you have a certain version installed. There is no way the first code snippet would have noticed it and let the failed, the second would have. Yes, this is whitebox and not blackbox testing but its a more meaningful test. 

But there is still a problem with just testing the body value of the mocked httpclient, all lines underneath [`resp, err := webhook.HTTPClient.Do(request)`](https://github.com/haveachin/infrared/pull/83/files#diff-71e3f02004a234858e10c42bf73e303ba86e5000ad4a1a95ccf5f6d94447c5e8R69) wont be tested. Also not sure how accurate the memory leak is, if it isnt being used the response body and the connection is closed the GC should collect the body of the response at some point and thus not leaking anything?

Reply:
I tried to clarify my intension with closing the body of the request. I also removed the EventLog return type, due to it being only used for tests.